cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/adapter/AdapterService.java

```diff
@@ -106,6 +120,27 @@ protected void shutDown() throws Exception {
   }
 
   /**
+   * Deploy the given application template in the given namespace.
+   *
+   * @param namespace the namespace to deploy in
+   * @param templateName the name of the template to deploy
+   * @throws NotFoundException if the template was not found
+   * @throws IllegalArgumentException if the template is invalid
+   * @throws IOException if there was an error reading the template jar
+   * @throws TimeoutException if there was a timeout examining the template jar
+   */
+  public void deployTemplate(Id.Namespace namespace, String templateName)
+    throws NotFoundException, InterruptedException, ExecutionException, TimeoutException, IOException {
+    ApplicationTemplateInfo templateInfo = appTemplateInfos.get(templateName);
+    if (templateInfo == null) {
+      throw new NotFoundException("ApplicationTemplate", templateName);
+    }
+    // make sure we're up to date on template info
+    registerTemplates();
+    deployTemplate(namespace, templateInfo);
```

Comment:
currently it will not fail, and is leaving it to the user to do the right thing depending on their template and the changes they've made.  It has the same behavior as deployment of an application. For workers, the old code will still be running until they restart the adapter. For workflow, the new code will be used at the next scheduled run. I think we eventually want the behavior of workflow to be like that of workers.